File: src/main/java/xyz/nkomarn/composter/network/pipeline/InboundHandler.java
package xyz.nkomarn.composter.network.pipeline;

import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.SimpleChannelInboundHandler;
import xyz.nkomarn.composter.network.Connection;
import xyz.nkomarn.composter.network.protocol.packet.serverbound.ServerboundPacket;

public class InboundHandler extends SimpleChannelInboundHandler<ServerboundPacket> {

    private final Connection connection;

    public InboundHandler(Connection connection) {
        this.connection = connection;
    }

    @Override
    protected void channelRead0(ChannelHandlerContext context, ServerboundPacket packet) throws Exception {
        System.out.println("Received packet " + packet.getClass().getSimpleName() + ".");
        packet.handle(connection.packetHandler());
    }

    @Override
    public void channelInactive(ChannelHandlerContext context) throws Exception {
        connection.handleDisconnection();
    }
}

File: src/main/kotlin/kyta/composter/network/pipeline/InboundHandler.kt
package kyta.composter.network.pipeline

import io.netty.channel.ChannelHandlerContext
import io.netty.channel.SimpleChannelInboundHandler
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kyta.composter.network.Connection
import kyta.composter.network.NetworkController
import kyta.composter.protocol.ServerboundPacket
import org.slf4j.Logger

class InboundHandler(
    private val connection: Connection,
    private val logger: Logger,
) : SimpleChannelInboundHandler<ServerboundPacket>() {
    override fun channelRead0(context: ChannelHandlerContext, packet: ServerboundPacket) {
        connection.launch {
            packet.handle(connection.packetHandler)
        }
    }

    override fun channelActive(context: ChannelHandlerContext) {
        logger.info("New session -> {}", context.channel().remoteAddress())
    }

    override fun channelInactive(context: ChannelHandlerContext) {
        connection.handleDisconnection()
        logger.info("Session closed <- {}", context.channel().remoteAddress())
    }

    override fun exceptionCaught(context: ChannelHandlerContext, cause: Throwable) {
        logger.error("an error occurred in ${connection.player.username}'s connection", cause)
    }
}